File: rnz_cropwise_flutter/lib/models/expert_request.dart
class ExpertRequest {
  final String? id;
  final String userId;
  final String? userEmail;
  final String? userName;
  final String plantType;
  final bool isHealthy;
  final String? diseaseName;
  final String? diseaseSeverity;
  final double confidenceScore;
  final String plantImage;
  final Map<String, dynamic> diagnosisResult;
  final String status;
  final String? adminNotes;
  final String? adminResponse;
  final DateTime? createdAt;
  final DateTime? updatedAt;
  final DateTime? respondedAt;
  final String? respondedBy;

  ExpertRequest({
    this.id,
    required this.userId,
    this.userEmail,
    this.userName,
    required this.plantType,
    required this.isHealthy,
    this.diseaseName,
    this.diseaseSeverity,
    required this.confidenceScore,
    required this.plantImage,
    required this.diagnosisResult,
    required this.status,
    this.adminNotes,
    this.adminResponse,
    this.createdAt,
    this.updatedAt,
    this.respondedAt,
    this.respondedBy,
  });

  factory ExpertRequest.fromJson(Map<String, dynamic> json) {
    return ExpertRequest(
      id: json['_id'],
      userId: json['userId'] ?? '',
      userEmail: json['userEmail'],
      userName: json['userName'],
      plantType: json['plantType'] ?? '',
      isHealthy: json['isHealthy'] ?? false,
      diseaseName: json['diseaseName'],
      diseaseSeverity: json['diseaseSeverity'],
      confidenceScore: (json['confidenceScore'] ?? 0.0).toDouble(),
      plantImage: json['plantImage'] ?? '',
      diagnosisResult: json['diagnosisResult'] ?? {},
      status: json['status'] ?? 'pending',
      adminNotes: json['adminNotes'],
      adminResponse: json['adminResponse'],
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
      respondedAt: json['respondedAt'] != null ? DateTime.parse(json['respondedAt']) : null,
      respondedBy: json['respondedBy'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (id != null) '_id': id,
      'userId': userId,
      if (userEmail != null) 'userEmail': userEmail,
      if (userName != null) 'userName': userName,
      'plantType': plantType,
      'isHealthy': isHealthy,
      if (diseaseName != null) 'diseaseName': diseaseName,
      if (diseaseSeverity != null) 'diseaseSeverity': diseaseSeverity,
      'confidenceScore': confidenceScore,
      'plantImage': plantImage,
      'diagnosisResult': diagnosisResult,
      'status': status,
      if (adminNotes != null) 'adminNotes': adminNotes,
      if (adminResponse != null) 'adminResponse': adminResponse,
      if (createdAt != null) 'createdAt': createdAt!.toIso8601String(),
      if (updatedAt != null) 'updatedAt': updatedAt!.toIso8601String(),
      if (respondedAt != null) 'respondedAt': respondedAt!.toIso8601String(),
      if (respondedBy != null) 'respondedBy': respondedBy,
    };
  }

  ExpertRequest copyWith({
    String? id,
    String? userId,
    String? userEmail,
    String? userName,
    String? plantType,
    bool? isHealthy,
    String? diseaseName,
    String? diseaseSeverity,
    double? confidenceScore,
    String? plantImage,
    Map<String, dynamic>? diagnosisResult,
    String? status,
    String? adminNotes,
    String? adminResponse,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? respondedAt,
    String? respondedBy,
  }) {
    return ExpertRequest(
      id: id ?? this.id,
      userId: userId ?? this.userId,
      userEmail: userEmail ?? this.userEmail,
      userName: userName ?? this.userName,
      plantType: plantType ?? this.plantType,
      isHealthy: isHealthy ?? this.isHealthy,
      diseaseName: diseaseName ?? this.diseaseName,
      diseaseSeverity: diseaseSeverity ?? this.diseaseSeverity,
      confidenceScore: confidenceScore ?? this.confidenceScore,
      plantImage: plantImage ?? this.plantImage,
      diagnosisResult: diagnosisResult ?? this.diagnosisResult,
      status: status ?? this.status,
      adminNotes: adminNotes ?? this.adminNotes,
      adminResponse: adminResponse ?? this.adminResponse,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      respondedAt: respondedAt ?? this.respondedAt,
      respondedBy: respondedBy ?? this.respondedBy,
    );
  }

  String get statusDisplay {
    switch (status) {
      case 'pending':
        return 'Pending';
      case 'in_progress':
        return 'In Progress';
      case 'completed':
        return 'Completed';
      case 'cancelled':
        return 'Cancelled';
      default:
        return 'Unknown';
    }
  }

  bool get isCompleted => status == 'completed';
  bool get isPending => status == 'pending';
  bool get isInProgress => status == 'in_progress';
  bool get isCancelled => status == 'cancelled';
}

File: src/lib/mongodb/models/ExpertRequest.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IExpertRequest extends Document {
  userId: string;
  userEmail?: string;
  userName?: string;
  plantType: string;
  isHealthy: boolean;
  diseaseName?: string;
  diseaseSeverity?: string;
  confidenceScore: number;
  plantImage: string; // Base64 image data
  diagnosisResult: any; // Full diagnosis result from AI
  status: 'pending' | 'in_progress' | 'completed' | 'cancelled';
  adminNotes?: string;
  adminResponse?: string;
  createdAt: Date;
  updatedAt: Date;
  respondedAt?: Date;
  respondedBy?: string; // Admin ID who responded
}

const ExpertRequestSchema = new Schema<IExpertRequest>({
  userId: {
    type: String,
    required: true,
    index: true
  },
  userEmail: {
    type: String,
    required: false
  },
  userName: {
    type: String,
    required: false
  },
  plantType: {
    type: String,
    required: true
  },
  isHealthy: {
    type: Boolean,
    required: true
  },
  diseaseName: {
    type: String,
    required: false
  },
  diseaseSeverity: {
    type: String,
    required: false
  },
  confidenceScore: {
    type: Number,
    required: true,
    min: 0,
    max: 1
  },
  plantImage: {
    type: String,
    required: true
  },
  diagnosisResult: {
    type: Schema.Types.Mixed,
    required: true
  },
  status: {
    type: String,
    enum: ['pending', 'in_progress', 'completed', 'cancelled'],
    default: 'pending',
    index: true
  },
  adminNotes: {
    type: String,
    required: false
  },
  adminResponse: {
    type: String,
    required: false
  },
  respondedAt: {
    type: Date,
    required: false
  },
  respondedBy: {
    type: String,
    required: false
  }
}, {
  timestamps: true
});

// Index for efficient queries
ExpertRequestSchema.index({ status: 1, createdAt: -1 });
ExpertRequestSchema.index({ userId: 1, createdAt: -1 });

export const ExpertRequest = mongoose.models.ExpertRequest || mongoose.model<IExpertRequest>('ExpertRequest', ExpertRequestSchema);
